### ISSUE TITLE
CLAHE Lacks Support for Other Image Types

### ISSUE THREAD

**Issue by opencv-pushbot** [Itseez/opencv#4770]
Transferred from http://code.opencv.org/issues/3749
```
|| Nathan Jackson on 2014-06-11 20:23
|| Priority: Normal
|| Affected: None
|| Category: imgproc, video
|| Tracker: Feature
|| Difficulty: 
|| PR: 
|| Platform: None / None
```

CLAHE Lacks Support for Other Image Types
-----------
```
Some of the data I work with is 16-bit grayscale.  I need to implement an algorithm that uses CLAHE, but the datatype is not supported.  From what I can tell, only 8-bit images are supported.  It would be extremely useful to be able to work with images of other datatypes than 8-bit grayscale.
```

History
-------
##### Nathan Jackson on 2014-06-18 18:45
```
I've begun work on this.  Currently working on getting 16-bit images to work.

See: https://github.com/nathanjackson/opencv/tree/clahe_datatypes
-   Assignee set to Nathan Jackson
```
##### Nathan Jackson on 2014-06-18 18:45
```
-   Status changed from New to Open
```
##### Daniil Osokin on 2014-07-07 16:41
```
Hi Nathan, looking forward to your contribution!
-   Category set to imgproc, video
```
##### Nathan Jackson on 2014-08-02 19:52
```
I've resumed work on this since my summer semester is over.
```
##### Nathan Jackson on 2014-08-03 00:22
```
In my work, I'm referencing some code from here: http://tog.acm.org/resources/GraphicsGems/gemsiv/clahe.c

I'm not sure if targetting 3.0-alpha is correct, someone may be able to better advise me on that.

16-bit images appear to be working and I've pushed the code into my repository.  I'm comparing my results using an ImageJ and a CLAHE plugin (http://rsbweb.nih.gov/ij/plugins/clahe/index.html).  The results from both implementations (OpenCV's and the ImageJ plugin) are the same.

Please feel free to test: https://github.com/nathanjackson/opencv/tree/clahe_datatypes

I want to further test the other datatypes.  I've only worked with 16-bit images so far as the data that I work with is 16-bit.
-   Target version set to 3.0-alpha
```
##### Nathan Jackson on 2014-08-03 23:02
```
A pull request has been made for the non-OpenCL version: https://github.com/Itseez/opencv/pull/3033
-   % Done changed from 0 to 30
```
##### Nathan Jackson on 2014-08-12 11:44
```
CLAHE now supports 16-bit images in mainline OpenCV, my pull request was merged.  I'm keeping this marked at 30% as I'd like to work on the OpenCL and CUDA implementations.

Only 16-bit images will be supported because the maintainers don't want to bloat the OpenCV binaries.
```
##### Maksim Shabunin on 2015-04-27 09:11
```
-   Target version changed from 3.0-alpha to 3.0
```

**Comment by dazzaphobus** [comment]
Testing on 16-bit images seems to produce very different output if I downscale to 8-bit first.

Comparing between the newly implemented CV_16UC1 version and your reference from GraphicGems, I believe this could be due to your lut being generated correctly for the histogram from source (srcPx >> 4) but the interpolate_body is applying the lut to the full 2^16 range.

The image looks a lot better (or more like 8-bit) if the output is shifted using the histogram bin size. I have also tested on changing clahe.cpp to test this.

**Comment by dazzaphobus** [comment]
Not sure if it's a bug, but testing on 16-bit images seems to produce very different output if I downscale to 8-bit first.

Comparing between the newly implemented CV_16UC1 version and your reference from GraphicGems, I believe this could be due to your lut being generated correctly for the histogram from source (srcPx >> 4) but the interpolate_body is applying the lut to the full 2^16 range.

The image looks a lot better (or more like 8-bit) if the output is shifted using the histogram bin size. I have also tested on changing clahe.cpp to test this and the output seems to be always the size of the histogram bin size which is incorrect.

**Comment by CelerityAbbottPC** [comment]
@nathanjackson : We just used the latest OpenCV 3.1 to process certain 16bit grayscale Tiff images, but the processed results looked very bad, and not acceptable. I can't to post my images here, but would send them to you in private. Could you see if you can process these few images 'correctly' to give satisfactory CLAHE results? Thanks in advance. (Same post as #6455)

**Comment by nathanjackson** [comment]
Sure, I can give it a shot. Just send them to me via the e-mail on my profile.

**Comment by dazzaphobus** [comment]
@nathanjackson I was using CLAHE awhile back and noticed the histogram bin size is 12bit (4096 steps) but can potentially be applied to a source image of 16-bit (rather than the expected 12-bit).
Perhaps @CelerityAbbottPC can try shifting left to 12-bit before performing CLAHE, and check if image is acceptable, but obviously some information will be lost in doing that, so better to check if it's an issue in the code itself.

**Comment by CelerityAbbottPC** [comment]
@nathanjackson , @dazzaphobus : You may know this already, but there is [a fix](https://github.com/Itseez/opencv/pull/6455/commits/43f854bc5f216d7e8dacc2144859a5a026c90dff) to some issue, and it basically did some bit wise shift to keep the 16bit data in 12bit.

(I always sent Nathan some testing images, and hope it helps.)

**Comment by dazzaphobus** [comment]
@CelerityAbbottPC The [fix](https://github.com/Itseez/opencv/pull/6455/commits/43f854bc5f216d7e8dacc2144859a5a026c90dff) will still produce different brightness output if it were to be processed as an 8-bit image.

You can test this by bit-shifting your 16-bit source into 8-bit before passing into CLAHE, and the output will be brighter than if you were to pass in your source as 16-bit.

In short, what you'll observe is:
CLAHE(16-bitSrc) != CLAHE(Cvt16to8(16-bitSrc))

I had to change the code to ensure the source images processed as 16-bit produces similar brightness as when I pass in the source image as 8-bit, except with higher dynamic range.

**Comment by CelerityAbbottPC** [comment]
@dazzaphobus : Thanks for the tip.  Would you be interested to receive my images, and use your modified code to process them and see?  I can send them to you in private.

**Comment by dazzaphobus** [comment]
@CelerityAbbottPC 
Yes. Please send through along with the parameter values you want. Email available on my profile. I hope I can be of some help

**Comment by pvtpeter** [comment]
Hi guys,

I encountered the same problem (~feature).
If there any pure solution for supporting 16-bit source in CLAHE?

@dazzaphobus: have you got any solution, modified code to handle CV_16UC1 images?

**Comment by nathanjackson** [comment]
@pvtpeter OpenCV should handle it now, this code was merged awhile ago. You're limited to a historgram bin size of 4096 for 16-bit images.

**Comment by pvtpeter** [comment]
@nathanjackson thanks for your reply. I know that opencv-master handle it right, but opencv-2.4.13 and opencv-3.1.0 crashed when i worked grayscale images which bit depth bigger than 12.
After that i tried openc-master and it worked without any crash, but the quality of images is objectionable.

If i normalize and convert the source 16 bit grayscale image to 8 bit and after that execute CLAHE, the result was beautiful, you can see it in this picture: 2.

If i execute only CLAHE on the source 16 bit grayscale image, the result was very weird. Little grids appeared on the image and it was very unnatural, , you can see it in this picture: 3.

I tried to different parametrization on CLAHE, but the point was the same.
I attached different pictures of my problem.

1. [Original 16 bit image normalize to 8 bit [0, 255] and convert to 8 bit (CV_8UC1).](http://kepfeltoltes.hu/161021/normalized8bitFromOriginalOutput_www.kepfeltoltes.hu_.jpg)
2. [Applying Clahe algorithm with default parameters (ClipLimit(40.0) and TilesGridSize(Size(8, 8))) on the previous (1.) image.](http://kepfeltoltes.hu/161021/claheOutput_www.kepfeltoltes.hu_.jpg)
3. [Applying Clahe algorithm with default parameters on the original 16 bit image. After executing, normalize Clahe result to 8 bit [0, 255] and convert to 8 bit (CV_8UC1).](http://kepfeltoltes.hu/161021/normalizedClaheOutput8Bit_www.kepfeltoltes.hu_.jpg)

I have got a question too: why is limited the histogram bin size to 4096 when a 16 bit image has got (1<<16)-1 == 65535 kinds shade?
And what do the grids mean on image 3? How can i eliminate them?

**Comment by nathanjackson** [comment]
@pvtpeter Its been a while since I've worked on this, so I'm trying to remember from a couple years ago.

So I was the first person to put a ticket in to address 16-bit images. However, the data I was working with at the time was actually 12-bit (stored as 16-bit values), so that's why 4096 was chosen, essentially because that's what I needed. It will probably not produce the best results for actual 16-bit data. To properly address this issue, you really need a more generic CLAHE implementation.

As far as your second question, I don't know for sure, but maybe it could have something to do with the 4096 bins maybe?

**Comment by pvtpeter** [comment]
@nathanjackson thanks again. I created an entry into OpenCV forum.
[Result of CLAHE is different on 8 and 16 bit grayscale image](http://answers.opencv.org/question/105472/result-of-clahe-is-different-on-8-and-16-bit-grayscale-image)
I hope this thread will solve the problem.

**Comment by jsxyhelu** [comment]
I really want to solove the problem,also.

**Comment by pvtpeter** [comment]
Hi Everybody,

Could someone tell me what causes the "living/moving" grids on the next 16 bit image?
http://i.imgur.com/SlCZAcb.gif (~17MB)
[Source *.png](https://drive.google.com/file/d/0BxYvuG9G0k78cElibmo3ek0tQXc)

I use the next parametrization:
ClipLimit is fix: 400.0
TileGridSize is various: (1,1) -> (2,2) -> (3,3) -> ... -> (56,56) -> (57,57)

BR,
Peter
